libopenfpga/libfpgaaesencryption/src/AESfunctions.cpp

```diff
@@ -0,0 +1,226 @@
+// https://en.wikipedia.org/wiki/Advanced_Encryption_Standard
+
+#include "AESfunctions.h"
+#include "AEStables.h"
+
+void KeyExpansionCore(unsigned char* in4, unsigned short i)
+{
+    // RotWord rotates left
+    // SubWord substitutes with S - Box value
+    unsigned char t = in4[0];
+    in4[0] = s_box[in4[1]];
+    in4[1] = s_box[in4[2]];
+    in4[2] = s_box[in4[3]];
+    in4[3] = s_box[t];
+    // RCon (round constant) 1st byte XOR rcon
+    in4[0] = in4[0] ^ rcon[i];
+}
+
+void SubWord(unsigned char* in4)
+{
+    // SubWord substitutes with S - Box value
+    in4[0] = s_box[in4[0]];
+    in4[1] = s_box[in4[1]];
+    in4[2] = s_box[in4[2]];
+    in4[3] = s_box[in4[3]];
+}
+
+void KeyExpansion(unsigned char* inputKey, unsigned short Nk, unsigned char* expandedKey)
+{
+    unsigned short Nr = (Nk > Nb) ? Nk + 6 : Nb + 6; // = 10, 12 or 14 rounds
+    // Copy the inputKey at the beginning of expandedKey
+    for (unsigned short i = 0; i < Nk * rows; i++) { expandedKey[i] = inputKey[i]; }
+
+    // Variables
+    unsigned short byGen = Nk * rows;
+    unsigned short rconIdx = 1;
+    unsigned char temp[rows];
+
+    // Generate expanded keys
+    while (byGen < (Nr + 1) * stt_lng)
+    {
+        // Read previously generated last 4 bytes (last word)
+        for (unsigned short i = 0; i < rows; i++) { temp[i] = expandedKey[byGen - rows + i]; }
+        // Perform KeyExpansionCore once for each 16 byte key
+        if (byGen % (Nk * rows) == 0)
+        {
+            KeyExpansionCore(temp, rconIdx);
+            rconIdx++;
+        }
+        else if ((Nk > 6) && (byGen % (Nk * rows) == (4 * rows)))
+        {
+            SubWord(temp);
+        }
+        // XOR temp with [bytesGenerated-16] and store in expandedKeys
+        for (unsigned short i = 0; i < rows; i++) {
+            expandedKey[byGen] = expandedKey[byGen - Nk * rows] ^ temp[i];
+            byGen++;
+        }
+    }
+}
+
+
+void SubBytes(unsigned char* state)
+{
+    for (unsigned short i = 0; i < stt_lng; i++) { state[i] = s_box[state[i]]; }
+}
+/*
+void InvSubBytes(unsigned char* state)
+{
+    for (unsigned short i = 0; i < stt_lng; i++) { state[i] = inverted_s_box[state[i]]; }
+}
+*/
+
+void ShiftRows(unsigned char* state)
+{
+    unsigned char tmp_state[stt_lng];
+    tmp_state[0] = state[0];
+    tmp_state[1] = state[5];
+    tmp_state[2] = state[10];
+    tmp_state[3] = state[15];
+
+    tmp_state[4] = state[4];
+    tmp_state[5] = state[9];
+    tmp_state[6] = state[14];
+    tmp_state[7] = state[3];
+
+    tmp_state[8] = state[8];
+    tmp_state[9] = state[13];
+    tmp_state[10] = state[2];
+    tmp_state[11] = state[7];
+
+    tmp_state[12] = state[12];
+    tmp_state[13] = state[1];
+    tmp_state[14] = state[6];
+    tmp_state[15] = state[11];
+    for (unsigned short i = 0; i < stt_lng; i++) { state[i] = tmp_state[i]; }
+}
+/*
```

Comment:
Is this function commented out by the authors or due to debugging?

Reply:
No I commented all the functions pertaining to the decipher code